runtime/pkg/graceful/grpc.go

```diff
@@ -25,17 +36,39 @@ func ServeGRPC(ctx context.Context, server *grpc.Server, port int) error {
 		return err
 	}
 
-	cctx, cancel := context.WithCancel(ctx)
-	var serveErr error
```

Comment:
I'm guessing the problem is contention around `serveErr`? I think this should be possible to handle with a single channel and without a mutex. What do you think about something like this?

```go
// Serve
serveErrCh := make(chan error)
go func() {
	err := server.Serve(lis)
	serveErrCh <- err
}

// Wait for cancellation or failure
select {
case err := <-serveErrCh:
	return err
case <-ctx.Done():
	server.GracefulStop()
}

// Wait for graceful shutdown
select {
case err := <-serveErrCh:
	return err
case <-time.After(grpcShutdownTimeout):
	server.Stop()
	return errors.New("grpc graceful shutdown timed out")
}
```

Reply:
Correct, I'll try this 👍